README.md
# Sensu Go Elasticsearch metric handler plugin
[![Bonsai Asset Badge](https://img.shields.io/badge/Sensu%20Go%20Elasticsearch-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/nixwiz/sensu-go-elasticsearch)[![Go Test](https://github.com/nixwiz/sensu-go-elasticsearch/workflows/Go%20Test/badge.svg)](https://github.com/nixwiz/sensu-go-elasticsearch/actions?query=workflow%3A%22Go+Test%22)[![Go Lint](https://github.com/nixwiz/sensu-go-elasticsearch/workflows/Go%20Lint/badge.svg)](https://github.com/nixwiz/sensu-go-elasticsearch/actions?query=workflow%3A%22Go+Lint%22)[![goreleaser](https://github.com/nixwiz/sensu-go-elasticsearch/workflows/goreleaser/badge.svg)](https://github.com/nixwiz/sensu-go-elasticsearch/actions?query=workflow%3A%22goreleaser%22)

## Table of Contents
- [Overview](#overview)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Handler definition](#handler-definition)
  - [Proxy support](#proxy-support)
  - [Annotations](#annotations)
- [Installation from source](#installation-from-source)
- [Contributing](#contributing)

## Overview

sensu-go-elasticsearch is a [Sensu Handler][2] for pushing metric data and full
event bodies into Elasticsearch for visualization in Kibana.

## Usage Examples

Help:

```
The Sensu Go handler for metric and event logging in elasticsearch
Required:  Set the ELASTICSEARCH_URL env var with an appropriate connection url (https://user:pass@hostname:port)

Usage:
  sensu-go-elasticsearch [flags]
  sensu-go-elasticsearch [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -d, --dated_index                 Should the index have the current date postfixed? ie: metric_data-2019-06-27
  -f, --full_event_logging          send the full event body instead of isolating event metrics
  -h, --help                        help for sensu-go-elasticsearch
  -i, --index string                index to use
  -s, --insecure-skip-verify        skip TLS certificate verification (not recommended!)
  -p, --point_name_as_metric_name   use the entire point name as the metric name
  -t, --trusted-ca-file string      TLS CA certificate bundle in PEM format

Use "sensu-go-elasticsearch [command] --help" for more information about a command.
```

## Configuration

### Asset registration

[Sensu Assets][3] are the best way to make use of this plugin. If you're not using an asset, please
consider doing so! If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the
following command to add the asset:

```
sensuctl asset add nixwiz/sensu-go-elasticsearch
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index][4]

### Handler definition

#### For metrics only

```yml
---
type: Handler
api_version: core/v2
metadata:
  name: sensu-go-elasticsearch
  namespace: default
spec:
  command: sensu-go-elasticsearch -i sensu --dated_index
  type: pipe
  runtime_assets:
    - nixwiz/sensu-go-elasticsearch
  secrets:
    - name: ELASTICSEARCH_URL
      secret: elasticsearch_url
  filters:
    - has_metrics
```

#### For full event data

```yml
---
type: Handler
api_version: core/v2
metadata:
  name: sensu-go-elasticsearch
  namespace: default
spec:
  command: sensu-go-elasticsearch -i sensu --dated_index --full_event_logging
  type: pipe
  runtime_assets:
    - nixwiz/sensu-go-elasticsearch
  secrets:
    - name: ELASTICSEARCH_URL
      secret: elasticsearch_url
```

### Environment Variables

The handler requires the ELASTICSEARCH_URL environment variable.  Given that the
username/password may be included in this URL, it is suggested to make use of
[secrets management][5] to surface it.  The handler definitions above reference
it as a secret.  Below is an example secret definition that makes use of the
built-in [env secrets provider][6].

```yml
---
type: Secret
api_version: secrets/v1
metadata:
  name: elasticsearch_url
spec:
  provider: env
  id: ELASTICSEARCH_URL
```

#### Proxy support

This handler supports the use of the environment variables HTTP_PROXY,
HTTPS_PROXY, and NO_PROXY (or the lowercase versions thereof). HTTPS_PROXY takes
precedence over HTTP_PROXY for https requests.  The environment values may be
either a complete URL or a "host[:port]", in which case the "http" scheme is assumed.

### Annotations

All arguments for this handler are tunable on a per entity or check basis based on annotations.  The
annotations keyspace for this handler is `sensu.io/plugins/elasticsearch/config`.

#### Examples

To change the index argument for a particular entity, in that entity's agent.yml add the following:

```yml
[...]
annotations:
  sensu.io/plugins/elasticsearch/config: "dev_index"
[...]
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset. If you would
like to compile and install the plugin from source or contribute to it, download the latest version
or create an executable from this source.

From the local path of the sensu-go-elasticsearch repository:

```
go build
```

## Contributing

For more information about contributing to this plugin, see [Contributing][1].

[1]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[2]: https://docs.sensu.io/sensu-go/latest/reference/handlers/
[3]: https://docs.sensu.io/sensu-go/latest/reference/assets/
[4]: https://bonsai.sensu.io/assets/nixwiz/sensu-go-elasticsearch
[5]: https://docs.sensu.io/sensu-go/latest/guides/secrets-management/
[6]: https://docs.sensu.io/sensu-go/latest/guides/secrets-management/#use-env-for-secrets-management



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

### Changed
- Minor README fixes

## [1.0.0] - 2020-06-20

### Changed
- Use new Sensu SDK
- Switch to GitHub Actions
- Updated README
- Add --insecure-skip-verify and --trusted-ca-file arguments

## [0.0.1] - 2000-01-01

### Added
- Initial release


lib/pkg/eventprocessing/eventprocessing.go
package eventprocessing

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
)

// MetricValue represents the format of metrics to send
type MetricValue struct {
	Timestamp string   `json:"timestamp"`
	Name      string   `json:"name"`
	Entity    string   `json:"entity"`
	Value     float64  `json:"value"`
	Namespace string   `json:"namespace"`
	Tags      []string `json:"tags"`
}

// EventValue represents the format of events to send
type EventValue struct {
	Timestamp string            `json:"timestamp"`
	Entity    *corev2.Entity    `json:"entity"`
	Check     *corev2.Check     `json:"check"`
	Metrics   *corev2.Metrics   `json:"namespace"`
	Metadata  corev2.ObjectMeta `json:"metadata"`
}

// {
// 	"name": "avg_cpu",
// 	"value": "56.0",
// 	"timestamp": "2019-03-30 12:30:00.45",
// 	"entity": "demo_test_agent",
// 	"namespace": "demo_jk185160",
// 	"tags": [
//    "company_jkte001",
//    "site_1001"
// 	]
// }

func parseTimestamp(timestamp int64) (string, error) {
	stringTimestamp := strconv.FormatInt(timestamp, 10)
	if len(stringTimestamp) > 10 {
		stringTimestamp = stringTimestamp[:10]
	}
	t, err := strconv.ParseInt(stringTimestamp, 10, 64)
	if err != nil {
		return "", err
	}
	return time.Unix(t, 0).Format(time.RFC3339), nil
}

func buildTag(key string, value string, prefix string) string {
	if len(prefix) > 0 {
		return fmt.Sprintf("%s_%s_%s", prefix, key, value)
	}
	return fmt.Sprintf("%s_%s", key, value)
}

// GetMetricFromPoint extracts the metrics points into a MetricValue type
func GetMetricFromPoint(point *corev2.MetricPoint, entityID string, namespaceID string, entityLabels map[string]string, pointNameAsMetricName bool) (MetricValue, error) {
	var metric MetricValue

	metric.Entity = entityID
	metric.Namespace = namespaceID
	// Find metric name
	if pointNameAsMetricName {
		metric.Name = point.Name
	} else {
		nameField := strings.Split(point.Name, ".")
		metric.Name = nameField[0]
	}
	// Find metric timstamp
	unixTimestamp, err := parseTimestamp(point.Timestamp)
	if err != nil {
		return *new(MetricValue), fmt.Errorf("failed to validate event: %s", err.Error())
	}
	metric.Timestamp = unixTimestamp
	metric.Tags = make([]string, len(point.Tags)+len(entityLabels)+1)
	i := 0
	for _, tag := range point.Tags {
		metric.Tags[i] = buildTag(tag.Name, tag.Value, "")
		i++
	}
	for key, val := range entityLabels {
		metric.Tags[i] = buildTag(key, val, "entity")
		i++
	}
	metric.Tags[i] = fmt.Sprintf("sensu_entity_name_%s", entityID)
	metric.Value = point.Value
	return metric, nil
}

// ParseEventTimestamp returns the event timestamp in a string format
func ParseEventTimestamp(event *corev2.Event) (EventValue, error) {
	var eventValue EventValue

	eventValue.Entity = event.Entity
	eventValue.Check = event.Check
	eventValue.Metrics = event.Metrics
	eventValue.Metadata = event.ObjectMeta

	timestamp, err := parseTimestamp(event.Timestamp)
	if err != nil {
		return *new(EventValue), fmt.Errorf("failed to validate event: %s", err.Error())
	}

	eventValue.Timestamp = timestamp
	return eventValue, nil
}


main.go
package main

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"time"

	"github.com/elastic/go-elasticsearch/v7"
	"github.com/elastic/go-elasticsearch/v7/esapi"
	"github.com/nixwiz/sensu-go-elasticsearch/lib/pkg/eventprocessing"
	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
)

// Config represents the handler plugin config.
type Config struct {
	sensu.PluginConfig
	DatedIndex            bool
	FullEventLogging      bool
	PointNameAsMetricName bool
	Index                 string
	TrustedCAFile         string
	InsecureSkipVerify    bool
}

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-go-elasticsearch",
			Short:    "The Sensu Go handler for metric and event logging in elasticsearch\nRequired:  Set the ELASTICSEARCH_URL env var with an appropriate connection url (https://user:pass@hostname:port)",
			Keyspace: "sensu.io/plugins/elasticsearch/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		{
			Path:      "dated_index",
			Env:       "",
			Argument:  "dated_index",
			Shorthand: "d",
			Default:   false,
			Usage:     "Should the index have the current date postfixed? ie: metric_data-2019-06-27",
			Value:     &plugin.DatedIndex,
		},
		{
			Path:      "full_event_logging",
			Env:       "",
			Argument:  "full_event_logging",
			Shorthand: "f",
			Default:   false,
			Usage:     "send the full event body instead of isolating event metrics",
			Value:     &plugin.FullEventLogging,
		},
		{
			Path:      "point_name_as_metric_name",
			Env:       "",
			Argument:  "point_name_as_metric_name",
			Shorthand: "p",
			Default:   false,
			Usage:     "use the entire point name as the metric name",
			Value:     &plugin.PointNameAsMetricName,
		},
		{
			Path:      "index",
			Env:       "",
			Argument:  "index",
			Shorthand: "i",
			Default:   "",
			Usage:     "index to use",
			Value:     &plugin.Index,
		},
		{
			Path:      "insecure-skip-verify",
			Env:       "",
			Argument:  "insecure-skip-verify",
			Shorthand: "s",
			Default:   false,
			Usage:     "skip TLS certificate verification (not recommended!)",
			Value:     &plugin.InsecureSkipVerify,
		},
		{
			Path:      "trusted-ca-file",
			Env:       "",
			Argument:  "trusted-ca-file",
			Shorthand: "t",
			Default:   "",
			Usage:     "TLS CA certificate bundle in PEM format",
			Value:     &plugin.TrustedCAFile,
		},
	}
)

func main() {
	handler := sensu.NewGoHandler(&plugin.PluginConfig, options, checkArgs, executeHandler)
	handler.Execute()
}

func checkArgs(event *corev2.Event) error {
	if len(plugin.Index) == 0 {
		return fmt.Errorf("No index specified")
	}
	return nil
}

func generateIndex() string {
	if plugin.DatedIndex {
		dt := time.Now()
		return fmt.Sprintf("%s-%s", plugin.Index, dt.Format("2006.01.02"))
	}
	return plugin.Index
}

func executeHandler(event *corev2.Event) error {
	if plugin.FullEventLogging {
		eventValue, err := eventprocessing.ParseEventTimestamp(event)
		if err != nil {
			return fmt.Errorf("error processing sensu event into eventValue: %v", err)
		}
		msg, err := json.Marshal(eventValue)
		if err != nil {
			return fmt.Errorf("error serializing metric data to json payload: %v", err)
		}
		err = sendElasticSearchData(string(msg), plugin.Index)
		if err != nil {
			return fmt.Errorf("error sending metric data to elasticsearch: %v", err)
		}
		return nil
	}

	if !event.HasMetrics() {
		return fmt.Errorf("event does not contain metrics")
	}
	for _, point := range event.Metrics.Points {
		metric, err := eventprocessing.GetMetricFromPoint(point, event.Entity.Name, event.Entity.Namespace, event.Entity.Labels, plugin.PointNameAsMetricName)
		if err != nil {
			return fmt.Errorf("error processing sensu event MetricPoints into MetricValue: %v", err)
		}
		msg, err := json.Marshal(metric)
		if err != nil {
			return fmt.Errorf("error serializing metric data to json payload: %v", err)
		}
		err = sendElasticSearchData(string(msg), plugin.Index)
		if err != nil {
			return fmt.Errorf("error sending metric data to elasticsearch: %v", err)
		}
	}
	return nil
}

func sendElasticSearchData(metricBody string, index string) error {
	var cfg elasticsearch.Config

	if len(plugin.TrustedCAFile) > 0 {
		cert, err := ioutil.ReadFile(plugin.TrustedCAFile)
		if err != nil {
			return err
		}
		cfg.CACert = cert
	} else if plugin.InsecureSkipVerify {
		cfg.Transport = &http.Transport{
			TLSClientConfig: &tls.Config{
				InsecureSkipVerify: true,
			},
		}
	}
	es, _ := elasticsearch.NewClient(cfg)
	req := esapi.IndexRequest{
		Index:   generateIndex(),
		Body:    strings.NewReader(metricBody),
		Refresh: "true",
	}

	// Perform the request with the client.
	res, err := req.Do(context.Background(), es)
	if err != nil {
		return fmt.Errorf("Error getting response: %s", err)
	}
	if res.IsError() {
		return fmt.Errorf("[%s] Error indexing document ID=%d", res.Status(), 0)
	}
	return nil
}
